### PR TITLE
move status check to manager setting

### PR DESCRIPTION
**Pull request by winterland1989**
respond to #193 , disable status check is very common in many places, so it should be part of manager setting. The only problem is that this patch is a breaking change. @snoyberg please judge on this.

### CODE REVIEW

**Comment by snoyberg** [comment]
I'm -1 on this as a breaking change if it can be avoided (which I think it can be). It's still quite useful to be able to decide on a per request basis whether or not to throw an exception on non-200s, and obviously letting existing code continue to function is a nice thing.

**Comment by winterland1989** [comment]
yes but i can't propose a good way with back compatibility to do this. in order to make old code continue to work, we have to add `checkStatus` field back with exactly same signature, and it's has to be able override manager setting(not other way around). but if we want to be able to express this overriding, the signature will have to be wrapped in a `Maybe`.

I'm happy to add back compatibility, but i need some input here.

**Comment by snoyberg** [comment]
I'll need to think about it a bit, let me get back to you next week.

File: http-client/Network/HTTP/Client/Cookies.hs [hunk 1]
```diff
@@ -47,7 +47,7 @@ isIpAddress =
             Just (i, x') | BS.null x' && i >= 0 && i < 256 -> go (rest - 1) y
             _ -> False
       where
-        (x, y') = BS.breakByte 46 bs -- period
+        (x, y') = BS.break (== 46) bs -- period
```
Comment (snoyberg) [inline http-client/Network/HTTP/Client/Cookies.hs]: Can these kinds of orthogonal tweaks be included in a separate PR?

**Comment by snoyberg** [comment]
Alright, let's start with the easier things to do to achieve future-compatibility when possible: instead of using accessors, we should be providing setter functions. So starting from the current version, we can add something like:

```haskell
setCheckStatus :: (Status -> ResponseHeaders -> CookieJar -> Maybe SomeException)
       -> Request -> Request
```

We can go further and deprecate the `checkStatus` accessor itself and recommend using `setCheckStatus`, which will make for a more graceful transition.

Now, as far as migrating towards having a manager-wide field with a request-level override. I can think of two approaches:

1. Change the type of `checkStatus` to be a `Maybe`, as you commented already. This is a breaking API change, but:
    * We retain all functionality of old versions
    * With the `setCheckStatus`, it will be trivial to write code that works with the old and new major version
2. A _really_ hacky trick: make the default value of `checkStatus` something like `\_ _ _ -> Just $ toException SpecialUnexportedTypeThatMeansUseTheDefaultHere`. That will technically allow us to express "please use the default from the manager" without any API change, though it will have a performance impact.

I'm not sure how important avoiding a breaking version number is here, and therefore whether we should do (2) or not. But one implementation strategy for all of this is:

1. Immediately add setters for all field accessors and deprecate the field accessors
2. Implement the hacky trick I mentioned
3. At some point in the future (after an announcement), we'll release a breaking API version of http-client that hides all of the field accessors and only exposes setters

**Comment by winterland1989** [comment]
I think the transition is a little bit complex, i'm not sure about the cost/value ratio now. what about keeping current behavior, and move http-conduit's `setRequestIgnoreStatus` to http-client? maybe rename to something short like `ignoreBadStatus`.

**Comment by snoyberg** [comment]
I'm fine with moving `setRequestIgnoreStatus` into http-client.

**Comment by winterland1989** [comment]
Ok, i will come up another pr, i want to get these questions clear before doing it:

+ if this name `ignoreBadStatus` ok to you?

+ do you want me to include the `breakByte/break` fix? it's cause warnings on my ghc, i can seperate it with a clear commit message.

**Comment by snoyberg** [comment]
I'd rather a name that is unambiguously a verb. The name suggested could be
taken as querying that ignore state.

I'm OK with the breakByte change, though I think the upstream warning was a
bad idea. I'd accept a PR for that change on its own, yes.

**Comment by winterland1989** [comment]
close in favor of #201
